File: hphp/hack/src/hhbc/ast_scope.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)
open Core_kernel
module T = Aast

module ScopeItem = struct
  type is_static = bool

  type is_async = bool

  type rx_level = Rx.t option

  type t =
    (* Named class *)
    | Class of Tast.class_
    (* Named function *)
    | Function of Tast.fun_
    (* Method in class *)
    | Method of Tast.method_
    (* PHP-style closure *)
    | LongLambda of is_static * is_async * rx_level
    (* Short lambda *)
    | Lambda of is_async * rx_level
end

module Scope = struct
  type t = ScopeItem.t list

  let toplevel : t = []

  let rec get_class scope =
    match scope with
    | [] -> None
    | ScopeItem.Class cd :: _ -> Some cd
    | _ :: scope -> get_class scope

  let rec get_span scope =
    match scope with
    | [] -> Pos.none
    | ScopeItem.Class cd :: _ -> cd.T.c_span
    | ScopeItem.Function fd :: _ -> fd.T.f_span
    | ScopeItem.Method md :: _ -> md.T.m_span
    | _ :: scope' -> get_span scope'

  let rec get_tparams scope =
    match scope with
    | [] -> []
    | ScopeItem.Class cd :: scope ->
      cd.T.c_tparams.T.c_tparam_list @ get_tparams scope
    | ScopeItem.Function fd :: scope -> fd.T.f_tparams @ get_tparams scope
    | ScopeItem.Method md :: scope -> md.T.m_tparams @ get_tparams scope
    | _ :: scope -> get_tparams scope

  let rec get_fun_tparams scope =
    match scope with
    | []
    | ScopeItem.Class _ :: _ ->
      []
    | ScopeItem.Function fd :: _ -> fd.T.f_tparams
    | ScopeItem.Method md :: _ -> md.T.m_tparams
    | _ :: scope -> get_fun_tparams scope

  let rec get_class_tparams scope =
    match scope with
    | [] -> { T.c_tparam_list = []; T.c_tparam_constraints = SMap.empty }
    | ScopeItem.Class cd :: _ -> cd.T.c_tparams
    | ScopeItem.Function _ :: scope
    | ScopeItem.Method _ :: scope
    | _ :: scope ->
      get_class_tparams scope

  let rec has_this scope =
    match scope with
    | [] -> true (* Assume top level has this *)
    | ScopeItem.Lambda _ :: scope
    | ScopeItem.LongLambda _ :: scope ->
      has_this scope
    | ScopeItem.Class _ :: _ -> false
    | ScopeItem.Function _ :: _ -> false
    | ScopeItem.Method _ :: _ -> true

  let rec is_in_async scope =
    match scope with
    | [] -> false
    | ScopeItem.Lambda _ :: scope
    | ScopeItem.LongLambda _ :: scope ->
      is_in_async scope
    | ScopeItem.Class _ :: _ -> false
    | ScopeItem.Function { T.f_fun_kind = kind; _ } :: _
    | ScopeItem.Method { T.m_fun_kind = kind; _ } :: _ ->
      kind = Ast_defs.FAsync || kind = Ast_defs.FAsyncGenerator

  let is_toplevel scope = scope = []

  let rec is_in_static_method scope =
    match scope with
    | ScopeItem.Method md :: _ -> md.T.m_static
    | ScopeItem.Lambda _ :: scope -> is_in_static_method scope
    | ScopeItem.LongLambda (is_static, _, _) :: scope ->
      (not is_static) && is_in_static_method scope
    | _ -> false

  let is_in_trait scope =
    match get_class scope with
    | None -> false
    | Some cd -> cd.T.c_kind = Ast_defs.Ctrait

  let is_in_lambda = function
    | ScopeItem.Lambda _ :: _
    | ScopeItem.LongLambda _ :: _ ->
      true
    | _ -> false

  let rec rx_of_scope scope =
    match scope with
    | []
    | ScopeItem.Class _ :: _ ->
      Rx.NonRx
    | ScopeItem.Lambda (_, Some rx_level) :: _
    | ScopeItem.LongLambda (_, _, Some rx_level) :: _ ->
      rx_level
    | ScopeItem.Lambda (_, None) :: rest
    | ScopeItem.LongLambda (_, _, None) :: rest ->
      rx_of_scope rest
    | ScopeItem.Method { T.m_user_attributes = attrs; _ } :: _
    | ScopeItem.Function { T.f_user_attributes = attrs; _ } :: _ ->
      Rx.rx_level_from_ast attrs |> Option.value ~default:Rx.NonRx

  let rec find_enclosing_function scope =
    match scope with
    | [] -> None
    | (ScopeItem.Function _ as f) :: _ -> Some f
    | (ScopeItem.Method _ as m) :: _ -> Some m
    | _ :: more -> find_enclosing_function more

  let find_function_attribute scope attr_name =
    let find_attribute ua_list =
      List.find ua_list (fun attr -> snd attr.T.ua_name = attr_name)
    in
    match find_enclosing_function scope with
    | Some (ScopeItem.Function f) -> find_attribute f.T.f_user_attributes
    | Some (ScopeItem.Method m) -> find_attribute m.T.m_user_attributes
    | _ -> None

  let has_function_attribute scope attr_name =
    Option.is_some @@ find_function_attribute scope attr_name
end

File: hphp/hack/src/hhbc/ast_scope.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

mod ast_scope_item;
pub use ast_scope_item::*;

use hhas_coeffects::HhasCoeffects;
use oxidized::{
    ast,
    ast_defs::{FunKind, Id},
    pos::Pos,
};

#[derive(Clone, Default, Debug)]
pub struct Scope<'a> {
    pub items: Vec<ScopeItem<'a>>,
}

impl<'a> Scope<'a> {
    pub fn toplevel() -> Self {
        Scope { items: vec![] }
    }

    pub fn push_item(&mut self, s: ScopeItem<'a>) {
        self.items.push(s)
    }

    pub fn iter(&self) -> impl ExactSizeIterator<Item = &ScopeItem<'a>> {
        self.items.iter().rev()
    }

    pub fn iter_subscopes(&self) -> impl Iterator<Item = &[ScopeItem<'a>]> {
        (0..self.items.len()).rev().map(move |i| &self.items[..i])
    }

    pub fn get_subscope_class<'b>(sub_scope: &'b [ScopeItem<'b>]) -> Option<&'b Class> {
        for scope_item in sub_scope.iter().rev() {
            if let ScopeItem::Class(cd) = scope_item {
                return Some(cd);
            }
        }
        None
    }

    pub fn get_class(&self) -> Option<&Class> {
        Self::get_subscope_class(&self.items[..])
    }

    pub fn get_span(&self) -> Pos {
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(cd) => {
                    return cd.get_span().clone();
                }
                ScopeItem::Function(fd) => {
                    return fd.get_span().clone();
                }
                ScopeItem::Method(md) => {
                    return md.get_span().clone();
                }
                _ => {}
            }
        }
        Pos::make_none()
    }

    pub fn get_tparams(&self) -> Vec<&ast::Tparam> {
        let mut tparams = vec![];
        let extend_shallowly = &mut |tps| {
            for tparam in tps {
                tparams.push(tparam);
            }
        };
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(cd) => {
                    extend_shallowly(cd.get_tparams());
                }
                ScopeItem::Function(fd) => {
                    extend_shallowly(fd.get_tparams());
                }
                ScopeItem::Method(md) => {
                    extend_shallowly(md.get_tparams());
                }
                _ => {}
            }
        }
        tparams
    }

    pub fn get_fun_tparams(&self) -> &[ast::Tparam] {
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(_) => {
                    return &[];
                }
                ScopeItem::Function(fd) => {
                    return fd.get_tparams();
                }
                ScopeItem::Method(md) => {
                    return md.get_tparams();
                }
                _ => {}
            }
        }
        &[]
    }

    pub fn get_class_tparams(&self) -> &[ast::Tparam] {
        for scope_item in self.iter() {
            if let ScopeItem::Class(cd) = scope_item {
                return cd.get_tparams();
            }
        }
        &[]
    }

    pub fn has_this(&self) -> bool {
        if self.items.is_empty() {
            /* Assume top level has this */
            return true;
        }
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(_) | ScopeItem::Function(_) => {
                    return false;
                }
                ScopeItem::Method(_) => {
                    return true;
                }
                _ => {}
            }
        }
        false
    }

    pub fn is_in_async(&self) -> bool {
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(_) => {
                    return false;
                }
                ScopeItem::Method(m) => {
                    let fun_kind = m.get_fun_kind();
                    return fun_kind == FunKind::FAsync || fun_kind == FunKind::FAsyncGenerator;
                }
                ScopeItem::Function(f) => {
                    let fun_kind = f.get_fun_kind();
                    return fun_kind == FunKind::FAsync || fun_kind == FunKind::FAsyncGenerator;
                }
                _ => {}
            }
        }
        false
    }

    pub fn is_toplevel(&self) -> bool {
        self.items.is_empty()
    }

    pub fn is_in_static_method(&self) -> bool {
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Method(md) => {
                    return md.is_static();
                }
                ScopeItem::LongLambda(_) => {}
                ScopeItem::Lambda(_) => {}
                _ => return false,
            }
        }
        false
    }

    pub fn is_in_lambda(&self) -> bool {
        self.items.last().map_or(false, &ScopeItem::is_in_lambda)
    }

    pub fn coeffects_of_scope(&self) -> HhasCoeffects {
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Class(_) => {
                    return HhasCoeffects::default();
                }
                ScopeItem::Method(m) => {
                    return HhasCoeffects::from_ast(m.get_ctxs(), m.get_params());
                }
                ScopeItem::Function(f) => {
                    return HhasCoeffects::from_ast(f.get_ctxs(), f.get_params());
                }
                ScopeItem::Lambda(Lambda { coeffects, .. })
                | ScopeItem::LongLambda(LongLambda { coeffects, .. })
                    if !coeffects.get_static_coeffects().is_empty() =>
                {
                    return coeffects.clone();
                }
                _ => {}
            }
        }
        HhasCoeffects::default()
    }

    pub fn has_function_attribute(&self, attr_name: impl AsRef<str>) -> bool {
        let has = |ua: &[ast::UserAttribute]| ua.iter().any(|a| a.name.1 == attr_name.as_ref());
        for scope_item in self.iter() {
            match scope_item {
                ScopeItem::Method(m) => {
                    return has(m.get_user_attributes());
                }
                ScopeItem::Function(f) => {
                    return has(f.get_user_attributes());
                }
                _ => {}
            }
        }
        false
    }

    pub fn is_static(&self) -> bool {
        for x in self.iter() {
            match x {
                ScopeItem::Function(_) => return true,
                ScopeItem::Method(md) => return md.is_static(),
                ScopeItem::LongLambda(_) => continue,
                ScopeItem::Lambda(_) => continue,
                _ => return true,
            }
        }
        true
    }

    // get captured variables when in closure scope
    pub fn get_captured_vars(&self) -> Vec<String> {
        // closure scope: lambda -> method -> class
        match &self.items[..] {
            [.., ScopeItem::Class(ast_cls), _, _] => ast_cls
                .get_vars()
                .iter()
                .map(|var| {
                    let Id(_, id) = &var.id;
                    format!("${}", id)
                })
                .collect::<Vec<_>>(),
            _ => panic!("closure scope should be lambda -> method -> class"),
        }
    }


    pub fn is_in_debugger_eval_fun(&self) -> bool {
        for x in self.iter() {
            match x {
                ScopeItem::LongLambda(_) | ScopeItem::Lambda(_) => continue,
                ScopeItem::Function(f) => return f.get_name().1 == "include",
                _ => return false,
            }
        }
        true
    }
}
